File: mobile/lib/core/models/notification.dart
/// Notification models for the Circlo safety app
/// 
/// Requirements: 8.1, 8.2, 8.3
/// - Push notifications via Firebase Cloud Messaging
/// - Encrypted notification payloads
/// - Priority-based notifications by circle type

/// Notification priority levels based on circle type
enum NotificationPriority {
  critical, // Inner Circle - highest priority
  high,     // Community Circle
  normal,   // Professional Circle
  low,      // General notifications
}

/// Notification delivery status
enum NotificationStatus {
  pending,   // Queued for delivery
  sent,      // Successfully sent to FCM
  delivered, // Confirmed delivered to device
  failed,    // Delivery failed
  expired,   // Notification expired before delivery
}

/// Types of notifications
enum NotificationType {
  alertCreated,
  alertVerified,
  alertEscalated,
  alertResolved,
  verificationRequest,
  checkRequest,
  circleInvite,
  circleUpdate,
  message,
  system,
}

/// Encrypted notification payload from FCM
class EncryptedNotificationPayload {
  final bool encrypted;
  final String payload;
  final String iv;
  final String notificationId;
  final String type;
  final String priority;
  final String timestamp;

  EncryptedNotificationPayload({
    required this.encrypted,
    required this.payload,
    required this.iv,
    required this.notificationId,
    required this.type,
    required this.priority,
    required this.timestamp,
  });

  factory EncryptedNotificationPayload.fromJson(Map<String, dynamic> json) {
    return EncryptedNotificationPayload(
      encrypted: json['encrypted'] as bool? ?? true,
      payload: json['payload'] as String,
      iv: json['iv'] as String,
      notificationId: json['notification_id'] as String,
      type: json['type'] as String,
      priority: json['priority'] as String,
      timestamp: json['timestamp'] as String,
    );
  }

  Map<String, dynamic> toJson() => {
    'encrypted': encrypted,
    'payload': payload,
    'iv': iv,
    'notification_id': notificationId,
    'type': type,
    'priority': priority,
    'timestamp': timestamp,
  };
}


/// Decrypted notification content
class NotificationContent {
  final String notificationId;
  final NotificationType type;
  final String title;
  final String body;
  final Map<String, dynamic>? data;
  final String? alertId;
  final String? circleId;
  final DateTime createdAt;

  NotificationContent({
    required this.notificationId,
    required this.type,
    required this.title,
    required this.body,
    this.data,
    this.alertId,
    this.circleId,
    required this.createdAt,
  });

  factory NotificationContent.fromJson(Map<String, dynamic> json) {
    return NotificationContent(
      notificationId: json['notification_id'] as String,
      type: _parseNotificationType(json['type'] as String),
      title: json['title'] as String,
      body: json['body'] as String,
      data: json['data'] as Map<String, dynamic>?,
      alertId: json['alert_id'] as String?,
      circleId: json['circle_id'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() => {
    'notification_id': notificationId,
    'type': type.name,
    'title': title,
    'body': body,
    'data': data,
    'alert_id': alertId,
    'circle_id': circleId,
    'created_at': createdAt.toIso8601String(),
  };

  static NotificationType _parseNotificationType(String type) {
    switch (type) {
      case 'alert_created':
        return NotificationType.alertCreated;
      case 'alert_verified':
        return NotificationType.alertVerified;
      case 'alert_escalated':
        return NotificationType.alertEscalated;
      case 'alert_resolved':
        return NotificationType.alertResolved;
      case 'verification_request':
        return NotificationType.verificationRequest;
      case 'check_request':
        return NotificationType.checkRequest;
      case 'circle_invite':
        return NotificationType.circleInvite;
      case 'circle_update':
        return NotificationType.circleUpdate;
      case 'message':
        return NotificationType.message;
      default:
        return NotificationType.system;
    }
  }
}

/// User notification preferences
class NotificationPreferences {
  final String id;
  final String userId;
  final bool innerCircleEnabled;
  final bool communityCircleEnabled;
  final bool professionalCircleEnabled;
  final bool alertNotifications;
  final bool messageNotifications;
  final bool circleNotifications;
  final bool systemNotifications;
  final bool quietHoursEnabled;
  final String? quietHoursStart;
  final String? quietHoursEnd;

  NotificationPreferences({
    required this.id,
    required this.userId,
    this.innerCircleEnabled = true,
    this.communityCircleEnabled = true,
    this.professionalCircleEnabled = true,
    this.alertNotifications = true,
    this.messageNotifications = true,
    this.circleNotifications = true,
    this.systemNotifications = true,
    this.quietHoursEnabled = false,
    this.quietHoursStart,
    this.quietHoursEnd,
  });

  factory NotificationPreferences.fromJson(Map<String, dynamic> json) {
    return NotificationPreferences(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      innerCircleEnabled: json['inner_circle_enabled'] as bool? ?? true,
      communityCircleEnabled: json['community_circle_enabled'] as bool? ?? true,
      professionalCircleEnabled: json['professional_circle_enabled'] as bool? ?? true,
      alertNotifications: json['alert_notifications'] as bool? ?? true,
      messageNotifications: json['message_notifications'] as bool? ?? true,
      circleNotifications: json['circle_notifications'] as bool? ?? true,
      systemNotifications: json['system_notifications'] as bool? ?? true,
      quietHoursEnabled: json['quiet_hours_enabled'] as bool? ?? false,
      quietHoursStart: json['quiet_hours_start'] as String?,
      quietHoursEnd: json['quiet_hours_end'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'user_id': userId,
    'inner_circle_enabled': innerCircleEnabled,
    'community_circle_enabled': communityCircleEnabled,
    'professional_circle_enabled': professionalCircleEnabled,
    'alert_notifications': alertNotifications,
    'message_notifications': messageNotifications,
    'circle_notifications': circleNotifications,
    'system_notifications': systemNotifications,
    'quiet_hours_enabled': quietHoursEnabled,
    'quiet_hours_start': quietHoursStart,
    'quiet_hours_end': quietHoursEnd,
  };

  NotificationPreferences copyWith({
    bool? innerCircleEnabled,
    bool? communityCircleEnabled,
    bool? professionalCircleEnabled,
    bool? alertNotifications,
    bool? messageNotifications,
    bool? circleNotifications,
    bool? systemNotifications,
    bool? quietHoursEnabled,
    String? quietHoursStart,
    String? quietHoursEnd,
  }) {
    return NotificationPreferences(
      id: id,
      userId: userId,
      innerCircleEnabled: innerCircleEnabled ?? this.innerCircleEnabled,
      communityCircleEnabled: communityCircleEnabled ?? this.communityCircleEnabled,
      professionalCircleEnabled: professionalCircleEnabled ?? this.professionalCircleEnabled,
      alertNotifications: alertNotifications ?? this.alertNotifications,
      messageNotifications: messageNotifications ?? this.messageNotifications,
      circleNotifications: circleNotifications ?? this.circleNotifications,
      systemNotifications: systemNotifications ?? this.systemNotifications,
      quietHoursEnabled: quietHoursEnabled ?? this.quietHoursEnabled,
      quietHoursStart: quietHoursStart ?? this.quietHoursStart,
      quietHoursEnd: quietHoursEnd ?? this.quietHoursEnd,
    );
  }
}

File: backend/app/models/notification.py
"""Notification database models.

Requirements: 8.1, 8.2, 8.3, 8.4
- Push notifications via Firebase Cloud Messaging
- Encrypted notification payloads
- Priority-based notifications by circle type
- Offline notification queuing
"""
import uuid
from datetime import datetime, timezone
from enum import Enum as PyEnum
from sqlalchemy import Column, String, DateTime, Boolean, ForeignKey, Enum, Text, Integer
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from ..core.database import Base


class NotificationPriority(str, PyEnum):
    """Notification priority levels based on circle type."""
    critical = "critical"  # Inner Circle - highest priority
    high = "high"          # Community Circle
    normal = "normal"      # Professional Circle
    low = "low"            # General notifications


class NotificationStatus(str, PyEnum):
    """Notification delivery status."""
    pending = "pending"    # Queued for delivery
    sent = "sent"          # Successfully sent to FCM
    delivered = "delivered"  # Confirmed delivered to device
    failed = "failed"      # Delivery failed
    expired = "expired"    # Notification expired before delivery


class NotificationType(str, PyEnum):
    """Types of notifications."""
    alert_created = "alert_created"
    alert_verified = "alert_verified"
    alert_escalated = "alert_escalated"
    alert_resolved = "alert_resolved"
    verification_request = "verification_request"
    check_request = "check_request"
    circle_invite = "circle_invite"
    circle_update = "circle_update"
    message = "message"
    system = "system"


class Notification(Base):
    """Notification model for push notifications."""
    
    __tablename__ = "notifications"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    
    # Notification content (encrypted)
    title_encrypted = Column(Text, nullable=False)
    body_encrypted = Column(Text, nullable=False)
    data_encrypted = Column(Text, nullable=True)  # Additional payload data
    iv = Column(String(32), nullable=False)  # Initialization vector for decryption
    
    # Notification metadata
    type = Column(Enum(NotificationType), nullable=False)
    priority = Column(Enum(NotificationPriority), default=NotificationPriority.normal)
    status = Column(Enum(NotificationStatus), default=NotificationStatus.pending)
    
    # Related entities
    alert_id = Column(UUID(as_uuid=True), ForeignKey("alerts.id", ondelete="SET NULL"), nullable=True, index=True)
    circle_id = Column(UUID(as_uuid=True), ForeignKey("circles.id", ondelete="SET NULL"), nullable=True)
    
    # Delivery tracking
    fcm_message_id = Column(String(255), nullable=True)  # Firebase message ID
    retry_count = Column(Integer, default=0)
    max_retries = Column(Integer, default=3)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    scheduled_at = Column(DateTime(timezone=True), nullable=True)  # For scheduled notifications
    sent_at = Column(DateTime(timezone=True), nullable=True)
    delivered_at = Column(DateTime(timezone=True), nullable=True)
    expires_at = Column(DateTime(timezone=True), nullable=True)  # Notification expiry
    
    # Relationships
    user = relationship("User", backref="notifications")
    alert = relationship("Alert", backref="notifications")
    circle = relationship("Circle", backref="notifications")


class DeviceToken(Base):
    """Device token model for FCM registration."""
    
    __tablename__ = "device_tokens"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    
    # FCM token
    token = Column(String(512), nullable=False, unique=True)
    platform = Column(String(20), nullable=False)  # 'ios', 'android', 'web'
    
    # Token status
    is_active = Column(Boolean, default=True)
    last_used_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))
    
    # Relationships
    user = relationship("User", backref="device_tokens")


class NotificationPreference(Base):
    """User notification preferences by circle type."""
    
    __tablename__ = "notification_preferences"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    
    # Preferences by circle type
    inner_circle_enabled = Column(Boolean, default=True)
    community_circle_enabled = Column(Boolean, default=True)
    professional_circle_enabled = Column(Boolean, default=True)
    
    # Notification type preferences
    alert_notifications = Column(Boolean, default=True)
    message_notifications = Column(Boolean, default=True)
    circle_notifications = Column(Boolean, default=True)
    system_notifications = Column(Boolean, default=True)
    
    # Quiet hours
    quiet_hours_enabled = Column(Boolean, default=False)
    quiet_hours_start = Column(String(5), nullable=True)  # HH:MM format
    quiet_hours_end = Column(String(5), nullable=True)    # HH:MM format
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))
    
    # Relationships
    user = relationship("User", backref="notification_preferences")
